### ISSUE TITLE
Support streaming of attachments

### ISSUE THREAD

**Issue by bschmalhofer** [RotherOSS/otobo#1397]
Currently the web server loads attachments into memory before delivering them as content. It would be nicer to do this in streaming mode.

**Comment by bschmalhofer** [comment]
The proposed change is to return an IO::Handle like object as the body. See https://metacpan.org/pod/PSGI#Body and https://stackoverflow.com/questions/6011793/psgi-response-what-kinds-of-filehandles-can-be-expected-to-work-with-psgi-and. This should work with nginx flawlessly, as the same approach is used for serving files in _var/httpd/htdocs_. For an example see https://github.com/plack/Plack/blob/master/lib/Plack/App/File.pm.

**Comment by bschmalhofer** [comment]
See also https://github.com/mojolicious/mojo/discussions/1884.

**Comment by bschmalhofer** [comment]
Full proxy support of the S3 web service is beyond the scope of this issue. But we only have four relevant cases:

- ArticleStorageDB: attachments are retrieved from DB, thus already in Memory => return as string
- ArticleStorageFS: attachment is file on file system => return as open file handle
- ArticleStorageS3 small file: asset is in memory => return as string
- ArticleStorageS3 large file: asset is a file => move file and return open file handle

First step is to make it work with ArticleStorageFS.

**Comment by bschmalhofer** [comment]
While looking into how to serve attachments as file handles, I found that for attachments the header 'Content-Length' was no longer added. There was only the incorrect comment:

>     # Content-Length will be added in Finalize()

This should be fixed using the function `Plack::Util::content_length()` in order to allow for file handles as response body.

**Comment by bschmalhofer** [comment]
TODO:

- [x] check for output filters that may be active for the frontend module AgentTicketAttachment
- [x] also implement streaming in the customer interface

**Comment by bschmalhofer** [comment]
A first test. using ArticleStorageS3, looked fine. Images could be uploaded and downloaded again. Files larger than 256 kB were synced to the file system before they were served a open file handles.

**Comment by bschmalhofer** [comment]
Instead of not returning a file handle when there is an output filter, I made the output filter support file handles. Tested in the Agent and in the Customer interface with the backends ArticleStorageFS and ArticleStorageS3. The inline and attached images could be retrieved and looked fine.
Encountered some errors when searching for customers, when adding a customer user. But this effect seems to be unrelated.

Not closing yet, as still some testing needs to be done.

- [x] further testing

**Comment by bschmalhofer** [comment]
While implementing support for serving attachments as file handles, I found that only few OTOBO pages include the Content-Length header. The cases I found were:

- inline and non-inline Attachments in the customer and agent interface
- AJAX responses
- responses from the generic interface

The reason, that Content-Length is rarely set, might be that in OTOBO 10.0.x the the Output was printed in chunks and thus the final length could not easily be determined. In OTOBO 10.1.x we always have that information as either a string or a real file handle is served.
Assuming that setting Content-Length is best practice I propose to:
- remove the instances where Content-Lenght is currently set
- activate https://metacpan.org/pod/Plack::Middleware::ContentLength
